File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic/BoundXmlProcessingInstruction.cs
using System.Diagnostics;
using Microsoft.CodeAnalysis.VisualBasic.Symbols;

namespace Microsoft.CodeAnalysis.VisualBasic
{
	internal sealed class BoundXmlProcessingInstruction : BoundExpression
	{
		private readonly BoundExpression _Target;

		private readonly BoundExpression _Data;

		private readonly BoundExpression _ObjectCreation;

		public BoundExpression Target => _Target;

		public BoundExpression Data => _Data;

		public BoundExpression ObjectCreation => _ObjectCreation;

		public BoundXmlProcessingInstruction(SyntaxNode syntax, BoundExpression target, BoundExpression data, BoundExpression objectCreation, TypeSymbol type, bool hasErrors = false)
			: base(BoundKind.XmlProcessingInstruction, syntax, type, hasErrors || BoundNodeExtensions.NonNullAndHasErrors(target) || BoundNodeExtensions.NonNullAndHasErrors(data) || BoundNodeExtensions.NonNullAndHasErrors(objectCreation))
		{
			_Target = target;
			_Data = data;
			_ObjectCreation = objectCreation;
		}

		[DebuggerStepThrough]
		public override BoundNode Accept(BoundTreeVisitor visitor)
		{
			return visitor.VisitXmlProcessingInstruction(this);
		}

		public BoundXmlProcessingInstruction Update(BoundExpression target, BoundExpression data, BoundExpression objectCreation, TypeSymbol type)
		{
			if (target != Target || data != Data || objectCreation != ObjectCreation || (object)type != base.Type)
			{
				BoundXmlProcessingInstruction boundXmlProcessingInstruction = new BoundXmlProcessingInstruction(base.Syntax, target, data, objectCreation, type, base.HasErrors);
				boundXmlProcessingInstruction.CopyAttributes(this);
				return boundXmlProcessingInstruction;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/BoundXmlProcessingInstruction.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports System
Imports System.Diagnostics

Namespace Microsoft.CodeAnalysis.VisualBasic
	Friend NotInheritable Class BoundXmlProcessingInstruction
		Inherits BoundExpression
		Private ReadOnly _Target As BoundExpression

		Private ReadOnly _Data As BoundExpression

		Private ReadOnly _ObjectCreation As BoundExpression

		Public ReadOnly Property Data As BoundExpression
			Get
				Return Me._Data
			End Get
		End Property

		Public ReadOnly Property ObjectCreation As BoundExpression
			Get
				Return Me._ObjectCreation
			End Get
		End Property

		Public ReadOnly Property Target As BoundExpression
			Get
				Return Me._Target
			End Get
		End Property

		Public Sub New(ByVal syntax As SyntaxNode, ByVal target As BoundExpression, ByVal data As BoundExpression, ByVal objectCreation As BoundExpression, ByVal type As TypeSymbol, Optional ByVal hasErrors As Boolean = False)
			MyBase.New(BoundKind.XmlProcessingInstruction, syntax, type, If(hasErrors OrElse target.NonNullAndHasErrors() OrElse data.NonNullAndHasErrors(), True, objectCreation.NonNullAndHasErrors()))
			Me._Target = target
			Me._Data = data
			Me._ObjectCreation = objectCreation
		End Sub

		<DebuggerStepThrough>
		Public Overrides Function Accept(ByVal visitor As BoundTreeVisitor) As BoundNode
			Return visitor.VisitXmlProcessingInstruction(Me)
		End Function

		Public Function Update(ByVal target As BoundExpression, ByVal data As BoundExpression, ByVal objectCreation As BoundExpression, ByVal type As TypeSymbol) As Microsoft.CodeAnalysis.VisualBasic.BoundXmlProcessingInstruction
			Dim boundXmlProcessingInstruction As Microsoft.CodeAnalysis.VisualBasic.BoundXmlProcessingInstruction
			If (target <> Me.Target OrElse data <> Me.Data OrElse objectCreation <> Me.ObjectCreation OrElse CObj(type) <> CObj(MyBase.Type)) Then
				Dim boundXmlProcessingInstruction1 As Microsoft.CodeAnalysis.VisualBasic.BoundXmlProcessingInstruction = New Microsoft.CodeAnalysis.VisualBasic.BoundXmlProcessingInstruction(MyBase.Syntax, target, data, objectCreation, type, MyBase.HasErrors)
				boundXmlProcessingInstruction1.CopyAttributes(Me)
				boundXmlProcessingInstruction = boundXmlProcessingInstruction1
			Else
				boundXmlProcessingInstruction = Me
			End If
			Return boundXmlProcessingInstruction
		End Function
	End Class
End Namespace